### PR TITLE
ember-data: Return a Promise from Model.save()

### PR DESCRIPTION
**Pull request by snewcomer**
Partial dup of https://github.com/emberjs/rfcs/pull/362#issuecomment-553646135 with improvements.  Specifically the addition of derived state to the promise wrapper.

[rendered](https://github.com/snewcomer/rfcs/blob/sn/model-save/text/0795-ember-data-return-promise-save.md)

### CODE REVIEW

**Comment by sandstrom** [comment]
Would it make more sense to return a vanilla Promise? 

Most of RSVP is already available in native Promise. 

(I'm guessing RSVP will eventually be deprecated too)

**Comment by snewcomer** [comment]
@sandstrom I like your suggestion. Might be something we want to RFC to get off of RSVP now that we are past IE11.  I'll talk to Chris tomorrow!

**Comment by snewcomer** [comment]
There are a few hairy parts where we were relying on private RSVP API's to guard against dangling promises.  Specifically [here](https://github.com/emberjs/data/blob/master/packages/store/addon/-private/system/store/finders.js#L6)...So alongside this PR, we are going to introduce a deprecation for RSVP.

File: text/0795-ember-data-return-promise-save.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+We do not believe this requires any update to the Ember curriculum. API documentation does not need to be updated as it is already documented as returning a promise. This behavior of accessing properties through the PromiseProxyMixin specifically for DS.Model.save() is not documented in the guides.
+
+## Drawbacks
+
+For users relying on this behavior, they may have to refactor their code to access model properties and methods, especially if they were reaching for private APIs.
+
+## Alternatives
+
+The impact of not doing this prevents further changes in Ember Data.
```
Comment (runspired) [inline text/0795-ember-data-return-promise-save.md:68]: alternatives also include:

- returning just a promise
- returning a promise but also exposing a template/js helper in a new package that wraps any promise exposing these flags providing broader utility without imposing this promise-state management on ember-data itself.

**Review by runspired — APPROVED** [review]

**Comment by runspired** [comment]
Added final comment period, planning on merging monday considering we previously moved this same RFC to FCP in #362

**Comment by jelhan** [comment]
> That said I would like to revisit the decision to add the state-flags. I think there's a strong case for it being a separate util, even if an official one.

I agree that an util function would be way better - especially for teaching. But I don't see how such an util function could be implemented. As far as I know there is no way to [synchronously determine a JavaScript Promise's state](https://stackoverflow.com/questions/30564053/how-can-i-synchronously-determine-a-javascript-promises-state).

**Comment by snewcomer** [comment]
@jelhan There has been a few approaches I have taken (the first being used with promise many arrays).

https://github.com/emberjs/data/pull/7868/files#diff-8aca4274066862134d43e99891338fd7bbd479fbb186cf0ab2f721fc55436a37

https://github.com/snewcomer/ember-stateful-promise/blob/main/addon/utils/stateful-promise.js

I'm unsure of what these two examples don't cover or if you were thinking about something else.  One downside to the util approach is for users to fish though their code to replace.  However, in conjunction with https://github.com/emberjs/rfcs/pull/796, I could see us not concerning ourselves with the derived state.

**Comment by jelhan** [comment]
> @jelhan There has been a few approaches I have taken (the first being pulled from promise many array implementation).
> 
> https://github.com/emberjs/data/pull/7868/files#diff-8aca4274066862134d43e99891338fd7bbd479fbb186cf0ab2f721fc55436a37
> 
> https://github.com/snewcomer/ember-stateful-promise/blob/main/addon/utils/stateful-promise.js
> 
> I'm unsure of what these two examples don't cover or if you were thinking about something else. One downside to the util approach is for users to fish though their code to replace. However, in conjunction with #796, I could see us not concerning ourselves with the derived state.

Thanks a lot for sharing this examples. I used the `ember-stateful-promise` library as an example to show the issue: https://github.com/snewcomer/ember-stateful-promise/pull/30

The state of a Promise, which has been resolved _before_ constructing a `StatefulPromise`, is reported wrongly as `isRunning` until next tick. As far as I know this limitation is by design and can not be fixed until ECMAScript provides an API to synchronously determine the state of a Promise.

I think that's the reason why @runspired proposes to have a `WeakMap` to provide a stable return for the same promise. That would resolve the issue of having two _different_ states reported for the same promise - depending if the stateful promise instance state was created before or after the Promise settled. But the state may still be wrong for one tick.

I fear that this API limitation would be very confusing to consumers. Let me illustrate using one _very_ confusing edge case based on the API proposed by @runspired in https://github.com/emberjs/rfcs/pull/795#issuecomment-1046318284:

```js
const promise = post.save();
await promise;

const state = promiseState(promise);
console.log('isPending', state.isPending); // true
console.log('isResolved', state.isResolved); // false

// do something async, which is totally unrelated,
// but causes later code to be executed in the next tick
await Promise.resolve();

console.log('isPending', state.isPending); // false
console.log('isResolved', state.isResolved); // true
```

:exploding_head:

**Comment by snewcomer** [comment]
@jelhan Really nice points.  Just a few notes...

1. The stateful-promise export is a promise, so I would expect it would be isRunning until the next tick.  A promise within a promise of sorts.

2. I can see how the latter example you provided **could** be an issue.  But in practice with async programming, more often than not, wouldn't be.

```
const statefulPromise = promiseState(promise);
await unrelated promise;
const result = await statefulPromise;

result // correct!
```

But I guess it would really trip people up if they used those state flags for non UI purposes.

```
const statefulPromise = promiseState(promise);
await unrelated promise;

if (statefulPromise.isRunning) {
  const result = await statefulPromise;
  result // correct!
}
```

So I agree with your take and runspireds.

**Comment by runspired** [comment]
@jelhan while it is true that potential for it to be wrong by one-tick exists, the WeakMap solution effectively solves that problem for the 99% case.

```ts
const promise = record.save();
const state = promiseState(promise);
// all state flags are correct before the await
await promise;
// all state flags are still correct after the await
// all state 2 flags are immediately correct, because state2 === state
const state2 = promiseState(promise);
```

 For the 1% case where someone fails to call promiseState until after the promise has resolved, there are two outcomes:
 
 1) the incorrect pending state is non-consequential since it gets fixed on the next tick (I suspect the majority of 1% cases would fall into this bucket)
 2) the incorrect pending state matters but is easily addressed by ensuring promiseState is called on the outcome since that was what was desired.

**Pull request merged**